### PR TITLE
Binary Heap implementation

### PR DESCRIPTION
**Pull request by bwetherfield**
For use as a priority queue, e.g. in Dijkstra's algorithm (shortest path).

### CODE REVIEW

File: Sources/DataStructures/BinaryHeap.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    // MARK: - Instance Properties
+    
+    private var storage: [Element]
+    private var lookup: [Element: Value]
+    private var indices: [Element: Int]
+    
+    // MARK: - Instance Methods
+    
+    private func lessAt(_ i: Int, than j: Int) -> Bool {
```
Comment (jsbean) [inline Sources/DataStructures/BinaryHeap.swift]: Move `private func` down to bottom of `struct` definition.

File: Sources/DataStructures/BinaryHeap.swift [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        return value(at: i) < value(at: j)
+    }
+    
+    mutating func insert (_ element: Element, _ value: Value) {
```
Comment (jsbean) [inline Sources/DataStructures/BinaryHeap.swift]: I don't dislike the whitespace here (`mutating func insert (`), but we should be consistent.

File: Sources/DataStructures/BinaryHeap.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                swapAt(i, j)
+                i = j
+            }
+            else { return }
+        }
+    }
+    
+    private mutating func balance () {
+        
+        func argmin(_ i: Int, _ j: Int) -> Int {
```
Comment (jsbean) [inline Sources/DataStructures/BinaryHeap.swift]: Is there a better name for this?

File: Sources/DataStructures/BinaryHeap.swift [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+    
+    mutating func insert (_ element: Element, _ value: Value) {
+        storage.append(element)
+        updateValue(of: element, to: value)
+        updateIndex(of: element, to: storage.count - 1)
+        bubbleUp(from: storage.count - 1)
+    }
+    
+    mutating func pop () -> (Element, Value)? {
```
Comment (jsbean) [inline Sources/DataStructures/BinaryHeap.swift]: Can we add some doc comments for any `public ` `func`, `var`, `init`, etc?

File: Sources/DataStructures/BinaryHeap.swift [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    
+    private mutating func bubbleUp (from i: Int) {
+        var i = i
+        while (i != 0) {
+            let j = (i-1)/2
+            if lessAt(i, than: j) {
+                swapAt(i, j)
+                i = j
+            }
+            else { return }
```
Comment (jsbean) [inline Sources/DataStructures/BinaryHeap.swift]: Elsewhere in the codebase, the cuddled else is used:

```Swift
if ... {
    ...
} else {
    ...
}
```

File: Tests/DataStructuresTests/BinaryHeapTests.swift [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        XCTAssertEqual(first.0, 2)
+        XCTAssertEqual(first.1, 0.5)
+        XCTAssertEqual(second.0, 1)
+        XCTAssertEqual(second.1, 3.5)
+        XCTAssertNil(heap.pop())
+    }
+    
+    func testBalance() {
+        var heap = BinaryHeap<Int, Double>()
+        var input: [(Int, Double)] = []
```
[the remark this answers is not in the document]
Comment (jsbean) [reply Tests/DataStructuresTests/BinaryHeapTests.swift]: Also, I think `input` could be modeled as `[Double]`.

Then, in order to get `element` later, you can do:

```Swift
for (element, value) in input.enumerated() {
    ...
}
```

[the remark this answers is not in the document]
Comment (jsbean) [reply Tests/DataStructuresTests/BinaryHeapTests.swift]: In this case, you don't pay for the space (as with storing `(Int,Double)`) (not that it is that critical…).

File: Sources/DataStructures/BinaryHeap.swift (hunk 3 above)
Comment (bwetherfield) [reply Sources/DataStructures/BinaryHeap.swift]: Not sure. It's a fairly standard term in the math world for the index at which the minimum is achieved, but will give it some thought

File: Sources/DataStructures/BinaryHeap.swift (hunk 2 above)
Comment (bwetherfield) [reply Sources/DataStructures/BinaryHeap.swift]: 👍

File: Sources/DataStructures/BinaryHeap.swift (hunk 1 above)
Comment (bwetherfield) [reply Sources/DataStructures/BinaryHeap.swift]: Good to know

File: Sources/DataStructures/BinaryHeap.swift (hunk 4 above)
Comment (bwetherfield) [reply Sources/DataStructures/BinaryHeap.swift]: For sure

File: Sources/DataStructures/BinaryHeap.swift (hunk 5 above)
Comment (bwetherfield) [reply Sources/DataStructures/BinaryHeap.swift]: Ah gotcha

File: Tests/DataStructuresTests/BinaryHeapTests.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (bwetherfield) [reply Tests/DataStructuresTests/BinaryHeapTests.swift]: Re the enumerated() thing. 💯 
The reason that I don't want it sorted is that I want to make sure the heap can handle arbitrary order of values. Do you think I should refactor that into a separate test?

File: Tests/DataStructuresTests/BinaryHeapTests.swift [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        for i in 0..<100 {
+            input.append( (i, Double.random(in: 0...1)) )
+        }
+        for (element, value) in input {
```
[the remark this answers is not in the document]
Comment (bwetherfield) [reply Tests/DataStructuresTests/BinaryHeapTests.swift]: nice

File: Tests/DataStructuresTests/BinaryHeapTests.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsbean) [reply Tests/DataStructuresTests/BinaryHeapTests.swift]: Ooh. Good point. Maybe that is better isolating that into a separate test, to prove a single semantic guarantee at a time.

File: Sources/DataStructures/BinaryHeap.swift (hunk 3 above)
Comment (bwetherfield) [reply Sources/DataStructures/BinaryHeap.swift]: Going with `minValueAt`

File: Tests/DataStructuresTests/BinaryHeapTests.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (bwetherfield) [reply Tests/DataStructuresTests/BinaryHeapTests.swift]: Hold on, in fact this is testing the balancing / the maintenance of binary heap invariant

File: Sources/DataStructures/BinaryHeap.swift [hunk 8]
```diff
@@ -0,0 +1,129 @@
+//
+//  BinaryHeap.swift
+//  PitchSpeller
+//
+//  Created by Benjamin Wetherfield on 7/15/18.
+//
+
+/// Implements a priority queue where minimum values have highest priority
+struct BinaryHeap<Element: Hashable, Value: Comparable> {
```
Comment (jsbean) [inline Sources/DataStructures/BinaryHeap.swift]: One last thing, can you add `Collection` conformance (you can do this just by forwarding a chain of `storage`, `lookup`, `indices` to `CollectionWrapping`):

```Swift
extension BinaryHeap: CollectionWrapping {
    var base: [Element] {
        ...
    }
}
```

Comment (bwetherfield) [reply Sources/DataStructures/BinaryHeap.swift]: Hold on, I'm not sure about this... Does this not break the interface of the heap structure? I could be missing something in the organization of the project. The `CollectionWrapping` is definitely a nice refactor!

Comment (jsbean) [reply Sources/DataStructures/BinaryHeap.swift]: Ah yes. You are totally right! Scratch this.

**Pull request merged**